### PR TITLE
adding Document and Documents Filters

### PR DESCRIPTION
**Pull request by AlexanderHaug**
These functions are rough but they are functional, planning to refactor further.

### CODE REVIEW

File: src/docs_filter.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+This program does the validation and assimilation of data for the documents jobs
+"""
+
+
+# Validation Functions
+def file_length_checker(json_data):
+    """
+    Checks the length of the file returned
+    :param json_data: path of the file to be checked
+    :return:
```
Comment (Charles-Peeke) [inline src/docs_filter.py]: Return values in function documentation

File: src/docs_filter.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        else:
+            file_count = 0
+            attachment_count = 0
+    return True
+
+
+# Assimilation Functions
+def documents_job(json_data, Redis_Manager):
+    """
+    Processes documents file and add ids to each document within
```
Comment (Charles-Peeke) [inline src/docs_filter.py]: "add ids to each document within...."? I think this was a partial thought or an incomplete sentence

File: src/docs_filter.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        job = create_job(workfile, job_id)
+        sff.add_job(job, Redis_Manager, "queued")
+
+
+def create_job(workfile, job_id):
+    """
+    Creates a job for the server to provide to clients
+    :param workfile: the job to be created and eventually completed
+    :param job_id: the id for the job
+    :return:
```
Comment (Charles-Peeke) [inline src/docs_filter.py]: Return value in documentation

File: src/docs_filter.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+
+# Final Function
+def process_docs(tup, Redis_Manager):
+    """
+    Main docs function, called by the server to compile list of document jobs
+    :param json_data: the json data for the jobs
+    :param Redis_Manager: database manager
+    :return:
+    """
```
Comment (Charles-Peeke) [inline src/docs_filter.py]: Extra spacing in this function

File: src/docs_filter.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    else:
+        if file_length_checker(json_data) and json_data["job_type"] == "documents":
+            documents_job(json_data, Redis_Manager)
+            key = sff.get_key_hash(json_data["job_id"], Redis_Manager, "progress")
+            sff.remove_job_progress(key, Redis_Manager, "progress")
+
+        else:
+            sff.renew_job(json_data, Redis_Manager)
+
+
```
Comment (Charles-Peeke) [inline src/docs_filter.py]: `PEP8` ... One new line at the end of a document

File: src/redis_manager.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import redis
+import threading
+from ast import literal_eval
+
+
```
Comment (Charles-Peeke) [inline src/redis_manager.py]: Spacing in general

File: src/redis_manager.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        self.lock = threading.RLock()
+
+    def get_work(self):
+        with self.lock:
+            work = literal_eval(self.r.lpop("queued").decode("utf-8"))
+            self.r.lpush("progress", work)
+            return work
+
+    def get(self, name):
+        """
```
Comment (Charles-Peeke) [inline src/redis_manager.py]: Most of this will be taken care of in another pull request

File: src/shared_filter_functions.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
+import json
+
```
Comment (dstocker) [inline src/shared_filter_functions.py]: This file should be integrated into the redis manager class

File: src/shared_filter_functions.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+# General Functions
+def add_job(job, Redis_Manager, queue):
```
Comment (Charles-Peeke) [inline src/shared_filter_functions.py]: I believe Daniel touched on this. We could integrate this file and functions into Redis Manager

File: src/shared_filter_functions.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    Redis_Manager.rpush(queue, job)
+
+
+def get_job(job_id, Redis_Manager, queue):
+    """
+    Selects a job from the queue by its id
+    :param job_id: the id for the job in question
+    :param Redis_Manager: database manager
+    :param queue: the queue containing jobs to be completed
+    :return:
```
Comment (Charles-Peeke) [inline src/shared_filter_functions.py]: Return value in function documentation

File: src/shared_filter_functions.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+
+# Queue Functions
+def job_exists(job_id, Redis_Manager, queue):
+    """
+    Verifies that a given job that is to be completed exists
+    :param job_id: the id for the job in question
+    :param Redis_Manager: database manager
+    :param queue: the queue containing jobs to be completed
+    :return:
```
Comment (Charles-Peeke) [inline src/shared_filter_functions.py]: Return value in function documentation

File: src/shared_filter_functions.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+    """
+    job = get_job(job_id, Redis_Manager, queue)
+    Redis_Manager.lrem(queue, 1, job)
+
+
+# Progress Functions
+def add_job_progress(job, Redis_Manager, queue, time):
+    """
+    TODO
```
Comment (Charles-Peeke) [inline src/shared_filter_functions.py]: TODO here... Probably the documentation

File: src/shared_filter_functions.py [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    :param queue:
+    :param time:
+    :return:
+    """
+    Redis_Manager.hset(queue, time, job)
+
+
+def job_exists_progress(key, Redis_Manager, queue):
+    """
+
```
Comment (Charles-Peeke) [inline src/shared_filter_functions.py]: Documentation

File: src/shared_filter_functions.py [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+    job = get_job_progress(key, Redis_Manager, queue)
+    if job == '':
+        return False
+    else:
+        return True
+
+
+def get_job_progress(key, Redis_Manager, queue):
+    """
+
```
Comment (Charles-Peeke) [inline src/shared_filter_functions.py]: Documentation

File: src/shared_filter_functions.py [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+    if job is None:
+        return ''
+    else:
+        data = job.decode("utf-8")
+        return data
+
+
+def get_key_hash(job_id, Redis_Manager, queue):
+    key_list = Redis_Manager.hgetall(queue)
```
Comment (Charles-Peeke) [inline src/shared_filter_functions.py]: Documentation

File: src/shared_filter_functions.py [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+    for key in key_list:
+        x = get_job_progress(key, Redis_Manager, queue)
+        info = json.loads(x)
+        if info["job_id"] == job_id:
+            return key.decode("utf-8")
+    return ''
+
+
+def remove_job_progress(key, Redis_Manager, queue):
+    Redis_Manager.hdel(queue, key)
```
Comment (Charles-Peeke) [inline src/shared_filter_functions.py]: Documentation

File: tests/test_doc_filter.py [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+
+
+def test_get_file_list_and_bad_work(savefile_tempdir):
+    compressed_file = PATH + "Bad_Archive.zip"
+    PATHstr = savefile_tempdir
+    file_list = df.get_file_list(compressed_file, PATHstr + "/")
+
+    condition = True
+    for file in file_list:
+        org, docket_id, document_id = df.get_doc_attributes(file)
```
Comment (Charles-Peeke) [inline tests/test_doc_filter.py]: Some extra spacing to be taken care of

File: tests/test_shared_filter_functions.py [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+
+
+def generate_data(job_id, type, data, version):
+    dict = {}
+
+    dict["job_id"] = job_id
+    dict["job_type"] = type
+    dict["data"] = data
+    dict["version"] = version
+
```
Comment (Charles-Peeke) [inline tests/test_shared_filter_functions.py]: Extra spaces

File: src/doc_filter.py [hunk 19]
```diff
@@ -0,0 +1,218 @@
+#!/usr/bin/env python
+import sys, os, os.path, tempfile, json, shutil, re, zipfile
+import shared_filter_functions as sff
```
Comment (dstocker) [inline src/doc_filter.py]: This will be changed to redis_manager

File: src/doc_filter.py [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+def process_doc(json_data, compressed_file, Redis_Manager):
+    """
+    Main doc function, called by the server to check and save documents returned from the client
+    :param json_data: json data of the job
+    :param compressed_file: compressed file of file names
+    :param Redis_Manager: database manager
+    :return:
+    """
+
+    if sff.job_exists_list(json_data, Redis_Manager, 'progress') is False:
```
Comment (dstocker) [inline src/doc_filter.py]: if r.does_job_exist_in_progress(job_id) is False:

File: src/doc_filter.py [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+                    break
+
+            elif file.startswith("doc.") and ending_is_number(document_id) and begining_is_letter(document_id):
+                    pass
+            else:
+                renew = True
+                break
+
+        if renew is True:
+            sff.renew_job(json_data, Redis_Manager)
```
Comment (dstocker) [inline src/doc_filter.py]: r.renew_job(job_id)

File: src/doc_filter.py [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+
+        else:
+            for file in file_list:
+                local_save(PATHstr + "/" + file, "~/regulations-data/")
+            key = sff.get_key_hash(json_data["job_id"], Redis_Manager, "progress")
```
Comment (dstocker) [inline src/doc_filter.py]: key = r.get_keys_from_progress(json_data["job_id"])

File: src/doc_filter.py [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+            sff.remove_job_hash(key, Redis_Manager, "progress")
```
Comment (dstocker) [inline src/doc_filter.py]: r.remove_job_from_progress(key)

File: src/docs_filter.py [hunk 24]
```diff
@@ -0,0 +1,79 @@
+import random
+import json
+import shared_filter_functions as sff
```
Comment (dstocker) [inline src/docs_filter.py]: redis_manager

File: src/docs_filter.py [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+    """
+    Creates a job for each workfile and then adds each to the "queue"
+    :param json_data: the json data containing all the workfiles
+    :param Redis_Manager: database manager
+    :return:
+    """
+    for workfile in json_data["data"]:
+        job_id = str(random.randint(0,1000000000))
+        job = create_document_job(workfile, job_id)
+        sff.add_job_list(job, Redis_Manager, "queue")
```
Comment (dstocker) [inline src/docs_filter.py]: r.add_to_queue(job)

File: src/docs_filter.py [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+# Final Function
+def process_docs(json_data, Redis_Manager):
+    """
+    Main documents function, called by the server to compile list of document jobs and add them to the "queue"
+    :param json_data: the json data for the jobs
+    :param Redis_Manager: database manager
+    :return:
+    """
+
+    if sff.job_exists_hash(json_data["job_id"], Redis_Manager, "progress") is False:
```
Comment (dstocker) [inline src/docs_filter.py]: if r.does_job_exist_in_progress(json_data["job_id"]) is False:

File: src/docs_filter.py [hunk 27]
```diff
@@ -0,0 +1,79 @@
...
+        pass
+
+    else:
+        if work_file_length_checker(json_data) and json_data["job_type"] == "documents":
+            add_document_job(json_data, Redis_Manager)
+            key = sff.get_key_hash(json_data["job_id"], Redis_Manager, "progress")
```
Comment (dstocker) [inline src/docs_filter.py]: r.get_keys_from_progres(json_data["job_id"])

File: src/docs_filter.py [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+            sff.remove_job_hash(key, Redis_Manager, "progress")
```
Comment (dstocker) [inline src/docs_filter.py]: r.remove_job_from_progress(key)

File: src/docs_filter.py [hunk 29]
```diff
@@ -0,0 +1,79 @@
...
+
+        else:
+            sff.renew_job(json_data, Redis_Manager)
```
Comment (dstocker) [inline src/docs_filter.py]: r.renew_job(job_id)

File: tests/test_docs_filter.py [hunk 30]
```diff
@@ -0,0 +1,63 @@
+import pytest
+import json
+import fakeredis
+
+import docs_filter as dsf
+import shared_filter_functions as sff
```
Comment (dstocker) [inline tests/test_docs_filter.py]: redis_manager

File: tests/test_docs_filter.py [hunk 31]
```diff
@@ -0,0 +1,63 @@
...
+    assert test_data["job_type"] == "documents"
+
+
+# Assimilation Tests
+def test_create_job():
+    r = setUp()
+    test_data = generate_json_data(PATH + '500_lines.json')
+    job_id = "1"
+    job = dsf.create_document_job(test_data["data"], job_id)
+    sff.add_job_list(job, r, "queue")
```
Comment (dstocker) [inline tests/test_docs_filter.py]: r.add_job_to_queue(job)

File: tests/test_docs_filter.py [hunk 32]
```diff
@@ -0,0 +1,63 @@
...
+    assert sff.job_exists_list('1', r, "queue")
```
Comment (dstocker) [inline tests/test_docs_filter.py]: r.does_job_exist_in_queue(job_id)

File: tests/test_shared_filter_functions.py [hunk 33]
```diff
@@ -0,0 +1,203 @@
+import pytest
+import fakeredis
+import json
+import time
+import shared_filter_functions as sff
```
Comment (dstocker) [inline tests/test_shared_filter_functions.py]: redis_manager

Comment (dstocker) [reply tests/test_shared_filter_functions.py]: Also ssf in this test file
